src/nvim/highlight_group.c

```diff
@@ -2901,6 +2901,387 @@ RgbValue name_to_color(const char *name, int *idx)
+    // resolve the highlight group link recursively
+    while (sgp->sg_link) {
+      hlgid = sgp->sg_link;
+      sgp = &hl_table[sgp->sg_link - 1];
+    }
+  }
+
+  if (sgp->sg_cterm != 0) {
+    attr_dict = highlight_get_attr_dict(sgp->sg_cterm);
+    if (attr_dict != NULL) {
+      if (tv_dict_add_dict(dict, S_LEN("cterm"), attr_dict) == FAIL) {
+        goto error;
+      }
+    }
+  }
+  if (sgp->sg_cterm_fg != 0) {
+    if (tv_dict_add_str(dict, S_LEN("ctermfg"),
+                        highlight_color(hlgid, (const char *)"fg", 'c')) == FAIL) {
+      goto error;
+    }
+  }
+  if (sgp->sg_cterm_bg != 0) {
+    if (tv_dict_add_str(dict, S_LEN("ctermbg"),
+                        highlight_color(hlgid, (const char *)"bg", 'c')) == FAIL) {
+      goto error;
+    }
+  }
+  if (sgp->sg_gui != 0) {
+    attr_dict = highlight_get_attr_dict(sgp->sg_gui);
+    if (attr_dict != NULL) {
+      if (tv_dict_add_dict(dict, S_LEN("gui"), attr_dict) == FAIL) {
+        goto error;
+      }
+    }
+  }
+  if (sgp->sg_rgb_fg_idx != kColorIdxNone) {
+    if (tv_dict_add_str(dict, S_LEN("guifg"),
+                        highlight_color(hlgid, (const char *)"fg", 'g')) == FAIL) {
+      goto error;
+    }
+  }
+  if (sgp->sg_rgb_bg_idx != kColorIdxNone) {
+    if (tv_dict_add_str(dict, S_LEN("guibg"),
+                        highlight_color(hlgid, (const char *)"bg", 'g')) == FAIL) {
+      goto error;
+    }
+  }
+  if (sgp->sg_rgb_sp_idx != kColorIdxNone) {
+    if (tv_dict_add_str(dict, S_LEN("guisp"),
+                        highlight_color(hlgid, (const char *)"sp", 'g')) == FAIL) {
+      goto error;
+    }
+  }
+  if (sgp->sg_link) {
+    char_u *link;
+
+    link = hl_table[sgp->sg_link - 1].sg_name;
+    if (link != NULL && tv_dict_add_str(dict, S_LEN("linksto"), (char *)link) == FAIL) {
+      goto error;
+    }
+  }
+  if (dict->dv_hashtab.ht_used == 2) {
+    // If only 'name' is present, then the highlight group is cleared.
+    tv_dict_add_bool(dict, S_LEN("cleared"), kBoolVarTrue);
+  }
+
+  return dict;
+
+error:
+  xfree(dict);
+  return NULL;
+}
+
+// "hlget([name])" function
+// Return the attributes of a specific highlight group (if specified) or all
+// the highlight groups.
+
+void f_hlget(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+{
+  list_T *list;
+  dict_T *dict;
+  int i;
+  const char *hlarg = NULL;
+  bool resolve_link = false;
+
+  tv_list_alloc_ret(rettv, kListLenMayKnow);
+
+  if (argvars[0].v_type != VAR_UNKNOWN) {
+    if (argvars[0].v_type != VAR_STRING) {
+      emsg("E1174: String required for argument 0");
+      return;
+    } else if (argvars[1].v_type != VAR_UNKNOWN
+               && (argvars[1].v_type != VAR_BOOL
+                   && !(argvars[1].v_type == VAR_NUMBER
+                        && (argvars[1].vval.v_number == 0
+                            || argvars[1].vval.v_number == 1)))) {
+      emsg("E1212: Bool required for argument 1");
+      return;
+    }
+  }
```

Comment:
Port the `tv_check_for_opt_string_arg()` and `tv_check_for_opt_bool_arg()` to `eval/typval.c`